pinot-core/src/main/java/com/linkedin/pinot/core/util/trace/Trace.java

```diff
@@ -17,79 +17,52 @@
 
 import java.util.ArrayList;
 import java.util.List;
-import java.util.concurrent.atomic.AtomicInteger;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.NotThreadSafe;
+
 
 /**
- * Trace maintains all trace logs for a specific Runnable or Callable job.
- *
- * It is not thread-safe, so should be protected by a {@link ThreadLocal}.
+ * Trace maintains all trace logs for a specific thread.
  */
-public class Trace {
-  private final AtomicInteger traceIdGen = new AtomicInteger(0);
+@NotThreadSafe
+public final class Trace {
+  private static class LogEntry {
+    private final String _key;
+    private final Object _value;
+
+    public LogEntry(String key, Object value) {
+      _key = key;
+      _value = value;
+    }
 
-  private String _traceId;
-  private List<String> _key = new ArrayList<String>();
-  private List<Object> _value = new ArrayList<Object>();
-  private Long _threadId;
-  protected final Trace _parent;
-  protected final List<Trace> _children = new ArrayList<Trace>();
+    @Override
+    public String toString() {
+      return _key + ": " + _value;
```

Comment:
Need to quote it to get the right json, otherwise we cannot parse it using standard json parsers.